tsdb: persist metric metadata to Parquet files

aknuds1 opened the pull request:
#### Which issue(s) does the PR fix:

Implementation of [proposal 3b](https://docs.google.com/document/d/1epBslSSwRO2do4armx40fruStJy_PS6thROnPeDifz8/edit?tab=t.0#heading=h.neppn3chuzf6) from the [Prometheus Metadata Store design doc](https://docs.google.com/document/d/1epBslSSwRO2do4armx40fruStJy_PS6thROnPeDifz8/edit). Fixes #12608.

#### What does this PR do:

Adds experimental Parquet-based series metadata persistence to TSDB, gated behind `--enable-feature=native-metadata`. Metric metadata (TYPE, HELP, UNIT) is stored in `series_metadata.parquet` sidecar files alongside TSDB blocks. Metadata is tracked per series (keyed by TSDB series ref) with full version history — each metadata change records the time range it was active. This feature is experimental and may change in future releases.

**API**
- `/api/v1/metadata`: when native metadata is enabled, uses TSDB as sole metadata source instead of scrape cache
- `/api/v1/targets/metadata`: queries TSDB head per-target via `SeriesMetadataForMatchers()`
- `/api/v1/metadata/versions`: _new_ endpoint returning time-versioned metadata history per series, with optional `metric`, `limit`, `start`, `end` parameters
- `/api/v1/metadata/series`: _new_ endpoint for inverse metadata lookup — find series by `type` (exact), `unit` (exact), or `help` (RE2 regex), with optional `limit`, `start`, `end` filtering

**Documentation**
- `docs/feature_flags.md`: feature flag description
- `tsdb/docs/format/series_metadata.md`: Parquet schema and merge semantics

**Demo**

A self-contained runnable demo is included in `documentation/examples/metadata-persistence/`. It starts a mock exporter, scrapes metrics with metadata into a temporary TSDB, exercises WAL replay, compacts to a Parquet sidecar file, and queries metadata back — demonstrating the full lifecycle. Run it with:

```bash
go run ./documentation/examples/metadata-persistence/...
```

#### Does this PR introduce a user-facing change?
```release-notes
[FEATURE] TSDB: Experimental support for persisting per-series versioned metric metadata (TYPE, HELP, UNIT) to Parquet sidecar files alongside TSDB blocks. Metadata is tracked per series with full version history, survives across compaction, and is available via `/api/v1/metadata` even after scrape targets are removed. Two new API endpoints are added: `/api/v1/metadata/versions` returns time-versioned metadata history per series, and `/api/v1/metadata/series` supports inverse metadata lookup (find series by type, unit, or help text). Enabled with `--enable-feature=native-metadata`. This feature is experimental and may change in future releases
```

#### TODO

- [x] Replicate functionality of Prometheus scrape metadata cache
- [x] Figure out whether we want to store metric metadata per target (currently flattened when storing to Parquet)
- [x] Implement the support for metric metadata changing over time, per time series, including API endpoint
- [x] Switch to TSDB references as foreign keys, instead of label set hash (to eliminate collision risk)
- [x] Support inverse lookup, i.e. find series with certain metadata
- [x] Review API endpoint response formats
- [ ] Move CLAUDE.md addition to separate PR

aknuds1 committed:
Metadata persisted to Parquet prototype

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

bboreham commented on documentation/examples/metadata-persistence/README.md:
````diff
@@ -0,0 +1,197 @@
+
+```
+┌─────────────────┐     scrape      ┌─────────────────┐
+│  Mock Exporter  │ ←───────────────│  Scrape/Parse   │
````
It strikes me as weird to specifcally call out "mock exporter" in the documentation.
Surely this is any exporter?

Should we show remote-write and OTLP also?

aknuds1 committed:
tsdb: replace scrape cache with TSDB as metadata source when native-metadata enabled

When --enable-feature=native-metadata is enabled, TSDB becomes the sole
metadata read source for all consumers instead of the scrape cache.

Key changes:
- seriesmetadata.Reader interface now supports multiple metadata entries
  per metric name (set semantics), reflecting that different series can
  report different {type, help, unit} for the same metric
- Add Head.SeriesMetadataForMatchers() for querying metadata by label
  matchers via the postings index
- Add DB.SeriesMetadataForMatchers() delegating to head, feature-gated
- /api/v1/metadata reads solely from TSDB when native-metadata enabled
- /api/v1/targets/metadata uses SeriesMetadataForMatchers with
  job+instance matchers per target, cached to avoid redundant queries
- Remote write MetadataWatcher reads from TSDB instead of scrape targets
- MetadataMetricsCollector reports TSDB-based metadata counts per job
- Metadata slices are sorted deterministically (by Type, Help, Unit)
  to ensure consistent Parquet output across compactions
- Set() with same hash but different metadata properly replaces the old
  entry in the byName set to prevent orphaned entries

The scrape cache continues as the write path (parsing TYPE/HELP/UNIT
into memSeries.meta). All behavior unchanged when feature is disabled.

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
tsdb: support time-varying metric metadata per series

Track metadata (type, unit, help) changes over time per series using
VersionedMetadata with AddOrExtend semantics. When metadata content
matches the latest version, the time range is extended; when it
changes, a new version is appended.

Persist versioned metadata to Parquet using a namespace-discriminated,
content-addressed schema with three namespaces: "metric" (backward
compat), "metadata_table" (deduplicated content), and
"metadata_mapping" (per-series version references).

Add MinTime/MaxTime fields to WAL RefMetadata records for timestamp
tracking. WAL replay uses AddOrExtend to preserve metadata history
through restarts.

Merge versioned metadata across blocks during compaction and
DB-level reads using MergeVersionedMetadata.

Expose /api/v1/metadata/versions endpoint with metric name filtering,
time range filtering, and limit support.

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
tsdb: use series refs as foreign keys in metadata Parquet files

Replace labelsHash (xxhash/64-bit) with TSDB series references
(HeadSeriesRef / BlockSeriesRef) as the foreign key in Parquet
metadata_mapping rows. This eliminates the theoretical risk of hash
collisions silently merging metadata from unrelated series.

In-memory maps are renamed from byHash/byHashVersioned to
byRef/byRefVersioned. The Parquet column is renamed from labels_hash
to series_ref. Head.SeriesMetadata() now uses s.ref directly,
removing the labels.StableHash() computation entirely.

Compaction and DB.SeriesMetadata() merges resolve refs to labels via
each source's index reader, merge by labels hash as a transient
in-memory key, then re-key with the destination block's refs.

Old-format Parquet files (with labels_hash column) will fail to load
and gracefully degrade to empty metadata, which is rebuilt on next
compaction. This is acceptable since native-metadata is feature-gated.

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
docs: update CLAUDE.md series metadata section for current branch state

Audit all bullets in the Series Metadata Persistence section against
source code and update stale descriptions: feature gate now documents
web.EnableNativeMetadata propagation, package bullet adds VersionedMetadata
and MetadataVersion core types, reader interface includes VersionedMetadataReader
methods, write path describes commitMetadata()/AddOrExtend flow and WAL replay,
merge path reflects labels.StableHash keying and SeriesMetadataForMatchers,
compaction documents 3-phase mergeAndWriteSeriesMetadata with ref resolution,
and web API covers all three endpoints and TSDB-backed metadata.

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
web: add /api/v1/metadata/series endpoint for inverse metadata lookup

Add a new endpoint that finds metrics whose metadata matches given
criteria, the inverse of /api/v1/metadata which looks up metadata for a
metric. Supports type (exact), unit (exact), and help (RE2 regex)
filters, with optional limit and start/end time range parameters.
Gated behind --enable-feature=native-metadata.

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
docs: update series metadata format doc for versioned schema

The Parquet schema has evolved to a two-table design with a namespace
discriminator, series refs as foreign keys, content-addressed metadata
tuples, and per-version time ranges. Update the format doc to reflect
the current schema and describe the compaction/merge behavior.

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Address reviewer feedback

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
web: return per-series metadata in versions and series endpoints

Change /api/v1/metadata/versions and /api/v1/metadata/series to return
per-series results with full label sets instead of merging by metric
name. This lets users trace metadata back to the originating series
(e.g. which job/instance had which help text).

The response format changes from map[metricName][]versionEntry to a
flat array of {labels, versions} objects. The limit parameter now
applies to series count rather than version entry count.

To support this, IterVersionedMetadata gains a labels.Labels parameter,
and the DB/head merge paths store labels via SetVersionedMetadataWithLabels.

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 commented:
Thanks for the feedback @beorn7!

> Implementing a backend for the existing metadata endpoints might appears as a good testing ground for a prototype, but I'd like to challenge that. First of all, the general agreement is that the existing metadata endpoints are not serving the purposes we need metadata for.

I wonder whether I've misunderstood @krajorama's feedback from our previous sync: "Try to replicate what the Prometheus scrape metadata cache does, this would be a good goal for this PR". I took that to mean that I should move this functionality from the scrape metadata cache to the new system, but maybe that's a misunderstanding on my part. @krajorama Could you clarify what you had in mind?

> My impression right now is that we mostly want a metadata store to support OTel entities, non-identifying OTel resource attributes (which theoretically do not exist, but in practice, the majority of resource attributes are non-identifying), and the necessary backend support to make the info function work as intended.

So, would you suggest that I drop the persisting of metric metadata from this prototype, and include support for OTel resource attributes instead (which I have in a follow-up branch)?

> Finally, I would like to understand better how the "reverse" lookup works efficiently. I don't know Parquet well enough to understand all the magic it provides. A quick glance at the code, however, appears to me as if an entries slice is materialized that has one element per series represented anywhere in the metadata (which would be … all series in the block, essentially, like hundreds of millions?). And then the whole slice is checked for matching entries. That's not just a "full table scan", it even puts everything into RAM at the same time. Maybe I'm missing something. Then an explanation would help…

I wouldn't look at the code right now. I'm currently just seeking high-level/directional feedback. If we agree on switching the prototype focus to OTel resource attributes though, I can pay more attention to the actual implementation.

krajorama commented:
> I wonder whether I've misunderstood @krajorama's feedback from our previous sync: "Try to replicate what the Prometheus scrape metadata cache does, this would be a good goal for this PR". I took that to mean that I should move this functionality from the scrape metadata cache to the new system, but maybe that's a misunderstanding on my part. @krajorama Could you clarify what you had in mind?

You understood perfectly, I thought that was a simple enough thing to start from and iterate on. But @beorn7 brings up excellent points. Just follow what he says.
